File: cpp/base/src/main/vsdk/toolkit/environment/material/MicroFacetedMaterial.cpp
#include <cctype>
#include <cstdio>
#include <cstdlib>

#include "java/lang/String.h"
#include "java/util/ArrayList.txx"
#include "vsdk/toolkit/environment/material/MicroFacetedMaterial.h"
struct MicrofacetConfig {
    java::String name;
    ColorRgb ambient;
    ColorRgb diffuse;
    ColorRgb specular;
    bool doubleSided;
    double reflectionCoefficient;
    double refractionCoefficient;
    double opacity;
    double phongExponent;
    double roughness;
    double alpha;
    ColorRgb fresnelF0;
    ColorRgb eta;
    ColorRgb kappa;
    double kd;
    double ks;
    int fresnelModel;
    int ndfModel;
    int geometryModel;
};

static java::String trim(const java::String& s)
{
    size_t a = 0;
    while ( a < s.size() && std::isspace((unsigned char)s[a]) ) a++;
    size_t b = s.size();
    while ( b > a && std::isspace((unsigned char)s[b - 1]) ) b--;
    return s.substr(a, b - a);
}

static java::String lower(const java::String& s)
{
    java::String out = s;
    for ( size_t i = 0; i < out.size(); i++ ) out[i] = (char)std::tolower((unsigned char)out[i]);
    return out;
}

static double clamp01(double v)
{
    if ( v < 0.0 ) return 0.0;
    if ( v > 1.0 ) return 1.0;
    return v;
}

static double parseDoubleOr(const java::String& s, double fallback)
{
    const char* str = trim(s).toCString();
    char* endptr = nullptr;
    double val = strtod(str, &endptr);
    if (endptr == str) return fallback;
    return val;
}

static int parseIntOr(const java::String& s, int fallback)
{
    const char* str = trim(s).toCString();
    char* endptr = nullptr;
    long val = strtol(str, &endptr, 10);
    if (endptr == str) return fallback;
    return (int)val;
}

static void splitCsv(const java::String& line, java::ArrayList<java::String>& cols)
{
    int prev = 0;
    for (int i = 0; i <= (int)line.size(); i++) {
        if (i == (int)line.size() || line[i] == ',') {
            cols.add(line.substr(prev, i - prev));
            prev = i + 1;
        }
    }
}

static MicrofacetConfig defaultConfig()
{
    MicrofacetConfig c;
    c.name = "Copper";
    c.ambient = ColorRgb(0.1, 0.1, 0.1);
    c.diffuse = ColorRgb(0.955008, 0.637427, 0.538163);
    c.specular = ColorRgb(0.955008, 0.637427, 0.538163);
    c.doubleSided = true;
    c.reflectionCoefficient = 0.0;
    c.refractionCoefficient = 0.0;
    c.opacity = 1.0;
    c.phongExponent = 40.0;
    c.roughness = 0.35;
    c.alpha = c.roughness * c.roughness;
    c.fresnelF0 = c.specular;
    c.eta = ColorRgb(0.2, 0.9, 1.1);
    c.kappa = ColorRgb(3.9, 2.5, 2.4);
    c.kd = 1.0;
    c.ks = 1.0;
    c.fresnelModel = MicroFacetedMaterial::FRESNEL_MODEL_SCHLICK;
    c.ndfModel = MicroFacetedMaterial::NDF_MODEL_BECKMANN;
    c.geometryModel = MicroFacetedMaterial::GEOMETRY_MODEL_SMITH;
    return c;
}

static MicrofacetConfig loadFromCsv(const java::String& csvFileName, const java::String& materialName)
{
    MicrofacetConfig d = defaultConfig();
    if ( csvFileName.empty() || materialName.empty() ) return d;

    FILE* in = fopen(csvFileName.c_str(), "r");
    if ( !in ) return d;

    char headerLineStr[4096];
    if ( !fgets(headerLineStr, sizeof(headerLineStr), in) ) { fclose(in); return d; }
    java::String headerLine(headerLineStr);
    java::ArrayList<java::String> header;
    splitCsv(headerLine, header);

    auto idx = [&](const java::String& n) -> int {
        const java::String needle = lower(n);
        for ( long int i = 0; i < header.size(); i++ ) {
            if ( lower(trim(header[i])) == needle ) return (int)i;
        }
        return -1;
    };

    const int iName = idx("material_name");
    const int iRoughness = idx("roughness");
    const int iAlpha = idx("alpha");
    const int iKd = idx("kd");
    const int iKs = idx("ks");
    const int iDr = idx("diffuse_r");
    const int iDg = idx("diffuse_g");
    const int iDb = idx("diffuse_b");
    const int iF0r = idx("f0_r");
    const int iF0g = idx("f0_g");
    const int iF0b = idx("f0_b");
    const int iEtaR = idx("eta_r");
    const int iEtaG = idx("eta_g");
    const int iEtaB = idx("eta_b");
    const int iKappaR = idx("kappa_r");
    const int iKappaG = idx("kappa_g");
    const int iKappaB = idx("kappa_b");
    const int iFModel = idx("fresnel_model");
    const int iNdf = idx("ndf_model");
    const int iGeo = idx("geometry_model");

    const java::String target = lower(trim(materialName));
    char lineStr[4096];
    while ( fgets(lineStr, sizeof(lineStr), in) ) {
        java::String line(lineStr);
        if ( trim(line).empty() ) continue;
        java::ArrayList<java::String> cols;
        splitCsv(line, cols);
        if ( iName < 0 || iName >= (int)cols.size() ) continue;
        const java::String rowName = trim(cols[(long int)iName]);
        if ( lower(rowName) != target ) continue;

        d.name = rowName;
        if ( iRoughness >= 0 && iRoughness < (int)cols.size() ) d.roughness = clamp01(parseDoubleOr(cols[(long int)iRoughness], d.roughness));
        if ( iAlpha >= 0 && iAlpha < (int)cols.size() ) d.alpha = clamp01(parseDoubleOr(cols[(long int)iAlpha], d.roughness * d.roughness));
        if ( iKd >= 0 && iKd < (int)cols.size() ) d.kd = clamp01(parseDoubleOr(cols[(long int)iKd], d.kd));
        if ( iKs >= 0 && iKs < (int)cols.size() ) d.ks = clamp01(parseDoubleOr(cols[(long int)iKs], d.ks));
        d.diffuse = ColorRgb(
            (iDr >= 0 && iDr < (int)cols.size()) ? parseDoubleOr(cols[(long int)iDr], d.diffuse.r()) : d.diffuse.r(),
            (iDg >= 0 && iDg < (int)cols.size()) ? parseDoubleOr(cols[(long int)iDg], d.diffuse.g()) : d.diffuse.g(),
            (iDb >= 0 && iDb < (int)cols.size()) ? parseDoubleOr(cols[(long int)iDb], d.diffuse.b()) : d.diffuse.b());
        d.fresnelF0 = ColorRgb(
            (iF0r >= 0 && iF0r < (int)cols.size()) ? parseDoubleOr(cols[(long int)iF0r], d.fresnelF0.r()) : d.fresnelF0.r(),
            (iF0g >= 0 && iF0g < (int)cols.size()) ? parseDoubleOr(cols[(long int)iF0g], d.fresnelF0.g()) : d.fresnelF0.g(),
            (iF0b >= 0 && iF0b < (int)cols.size()) ? parseDoubleOr(cols[(long int)iF0b], d.fresnelF0.b()) : d.fresnelF0.b());
        d.specular = d.fresnelF0;
        d.eta = ColorRgb(
            (iEtaR >= 0 && iEtaR < (int)cols.size()) ? parseDoubleOr(cols[(long int)iEtaR], d.eta.r()) : d.eta.r(),
            (iEtaG >= 0 && iEtaG < (int)cols.size()) ? parseDoubleOr(cols[(long int)iEtaG], d.eta.g()) : d.eta.g(),
            (iEtaB >= 0 && iEtaB < (int)cols.size()) ? parseDoubleOr(cols[(long int)iEtaB], d.eta.b()) : d.eta.b());
        d.kappa = ColorRgb(
            (iKappaR >= 0 && iKappaR < (int)cols.size()) ? parseDoubleOr(cols[(long int)iKappaR], d.kappa.r()) : d.kappa.r(),
            (iKappaG >= 0 && iKappaG < (int)cols.size()) ? parseDoubleOr(cols[(long int)iKappaG], d.kappa.g()) : d.kappa.g(),
            (iKappaB >= 0 && iKappaB < (int)cols.size()) ? parseDoubleOr(cols[(long int)iKappaB], d.kappa.b()) : d.kappa.b());
        if ( iFModel >= 0 && iFModel < (int)cols.size() ) d.fresnelModel = parseIntOr(cols[(long int)iFModel], d.fresnelModel);
        if ( iNdf >= 0 && iNdf < (int)cols.size() ) d.ndfModel = parseIntOr(cols[(long int)iNdf], d.ndfModel);
        if ( iGeo >= 0 && iGeo < (int)cols.size() ) d.geometryModel = parseIntOr(cols[(long int)iGeo], d.geometryModel);
        fclose(in);
        return d;
    }
    fclose(in);
    return d;
}
MicroFacetedMaterial::MicroFacetedMaterial()
    : SimpleMaterial(
        defaultConfig().name,
        defaultConfig().ambient,
        defaultConfig().diffuse,
        defaultConfig().specular,
        defaultConfig().doubleSided,
        defaultConfig().reflectionCoefficient,
        defaultConfig().refractionCoefficient,
        defaultConfig().opacity,
        defaultConfig().phongExponent),
      roughness(defaultConfig().roughness),
      alpha(defaultConfig().alpha),
      fresnelF0(defaultConfig().fresnelF0),
      eta(defaultConfig().eta),
      kappa(defaultConfig().kappa),
      kd(defaultConfig().kd),
      ks(defaultConfig().ks),
      fresnelModel(defaultConfig().fresnelModel),
      ndfModel(defaultConfig().ndfModel),
      geometryModel(defaultConfig().geometryModel)
{
}

MicroFacetedMaterial::MicroFacetedMaterial(const MicroFacetedMaterial& other)
    : SimpleMaterial(other),
      roughness(other.roughness),
      alpha(other.alpha),
      fresnelF0(other.fresnelF0),
      eta(other.eta),
      kappa(other.kappa),
      kd(other.kd),
      ks(other.ks),
      fresnelModel(other.fresnelModel),
      ndfModel(other.ndfModel),
      geometryModel(other.geometryModel)
{
}

MicroFacetedMaterial::MicroFacetedMaterial(const java::String& csvFileName, const java::String& materialName)
    : SimpleMaterial(),
      roughness(0.35),
      alpha(0.35 * 0.35),
      fresnelF0(1, 1, 1),
      eta(0.2, 0.9, 1.1),
      kappa(3.9, 2.5, 2.4),
      kd(1.0),
      ks(1.0),
      fresnelModel(FRESNEL_MODEL_SCHLICK),
      ndfModel(NDF_MODEL_BECKMANN),
      geometryModel(GEOMETRY_MODEL_SMITH)
{
    MicrofacetConfig c = loadFromCsv(csvFileName, materialName);
    *this = MicroFacetedMaterial();
    SimpleMaterial tmp(
        c.name, c.ambient, c.diffuse, c.specular, c.doubleSided,
        c.reflectionCoefficient, c.refractionCoefficient, c.opacity, c.phongExponent);
    *((SimpleMaterial*)this) = tmp;
    roughness = c.roughness;
    alpha = c.alpha;
    fresnelF0 = c.fresnelF0;
    eta = c.eta;
    kappa = c.kappa;
    kd = c.kd;
    ks = c.ks;
    fresnelModel = c.fresnelModel;
    ndfModel = c.ndfModel;
    geometryModel = c.geometryModel;
}

double MicroFacetedMaterial::getRoughness() const { return roughness; }
double MicroFacetedMaterial::getAlpha() const { return alpha; }
ColorRgb MicroFacetedMaterial::getFresnelF0() const { return ColorRgb(fresnelF0); }
ColorRgb MicroFacetedMaterial::getEta() const { return ColorRgb(eta); }
ColorRgb MicroFacetedMaterial::getKappa() const { return ColorRgb(kappa); }
double MicroFacetedMaterial::getKd() const { return kd; }
double MicroFacetedMaterial::getKs() const { return ks; }
int MicroFacetedMaterial::getFresnelModel() const { return fresnelModel; }
int MicroFacetedMaterial::getNdfModel() const { return ndfModel; }
int MicroFacetedMaterial::getGeometryModel() const { return geometryModel; }

File: typescript/base/src/main/vsdk/toolkit/environment/material/MicroFacetedMaterial.ts
import { ColorRgb } from "../../common/color/ColorRgb.js";
import { SimpleMaterial } from "./SimpleMaterial.js";
export class MicrofacetConfig {
    public constructor(
        public readonly name = "VSDK_default_material",
        public readonly ambient = new ColorRgb(0.1, 0.1, 0.1),
        public readonly diffuse = new ColorRgb(0.9, 0.5, 0.5),
        public readonly specular = new ColorRgb(1, 1, 1),
        public readonly doubleSided = true,
        public readonly reflectionCoefficient = 0,
        public readonly refractionCoefficient = 0,
        public readonly opacity = 1,
        public readonly phongExponent = 128,
        public readonly roughness = 0.35,
        public readonly alpha = 0.1225,
        public readonly fresnelF0 = new ColorRgb(0.04, 0.04, 0.04),
        public readonly eta = new ColorRgb(1.5, 1.5, 1.5),
        public readonly kappa = new ColorRgb(),
        public readonly kd = 1,
        public readonly ks = 1,
        public readonly fresnelModel = 0,
        public readonly ndfModel = 0,
        public readonly geometryModel = 0,
    ) {}
}
/**
The platform-neutral half of Java's `MicroFacetedMaterial(String csvFileName,
String materialName)` constructor.

Java's constructor does two things: it finds the CSV file — `resolveCsvFile`
tries the name as given and then under `etc/materials` — and it parses it. Only
the first is bound to a file system, so it stays with the caller, which hands
the already-read text over here; `@vitral/fs` reads it from disk and a browser
fetches it. The parse itself is the Java one, `readAsciiLine` by
`readAsciiLine`: the header row names the columns, a row matching the material
name case-insensitively supplies whatever fields it carries, and every absent
or blank field falls back to the same default the Java reader uses, which for
`alpha` is `roughness * roughness`.

Java's `readAsciiLine` throws at end of file, and the reading loop treats that
throw as the end of the table, so a final line with no terminating newline is
never seen. Splitting the text on newlines would see it, which is why the split
below drops a trailing empty segment only, leaving that Java behavior intact
for a file that ends properly and reproducing it for one that does not.
*/
export class MicroFacetedMaterial extends SimpleMaterial {
    public static readonly FRESNEL_MODEL_SCHLICK = 0;
    public static readonly FRESNEL_MODEL_CONDUCTOR = 1;
    public static readonly NDF_MODEL_BECKMANN = 0;
    public static readonly NDF_MODEL_GGX = 1;
    public static readonly GEOMETRY_MODEL_SMITH = 0;
    public static readonly GEOMETRY_MODEL_IMPLICIT = 1;
    private readonly config: MicrofacetConfig;
    public constructor();
    public constructor(other: MicroFacetedMaterial);
    public constructor(config: MicrofacetConfig);
    public constructor(value: MicroFacetedMaterial | MicrofacetConfig = new MicrofacetConfig()) {
        const c = value instanceof MicroFacetedMaterial ? value.config : value;
        super(
            c.name,
            c.ambient,
            c.diffuse,
            c.specular,
            c.doubleSided,
            c.reflectionCoefficient,
            c.refractionCoefficient,
            c.opacity,
            c.phongExponent,
        );
        this.config = new MicrofacetConfig(
            c.name,
            new ColorRgb(c.ambient),
            new ColorRgb(c.diffuse),
            new ColorRgb(c.specular),
            c.doubleSided,
            c.reflectionCoefficient,
            c.refractionCoefficient,
            c.opacity,
            c.phongExponent,
            Math.max(0, Math.min(1, c.roughness)),
            Math.max(0, Math.min(1, c.alpha)),
            new ColorRgb(c.fresnelF0),
            new ColorRgb(c.eta),
            new ColorRgb(c.kappa),
            Math.max(0, Math.min(1, c.kd)),
            Math.max(0, Math.min(1, c.ks)),
            c.fresnelModel,
            c.ndfModel,
            c.geometryModel,
        );
    }
    public getRoughness() {
        return this.config.roughness;
    }
    public getAlpha() {
        return this.config.alpha;
    }
    public getFresnelF0() {
        return new ColorRgb(this.config.fresnelF0);
    }
    public getEta() {
        return new ColorRgb(this.config.eta);
    }
    public getKappa() {
        return new ColorRgb(this.config.kappa);
    }
    public getKd() {
        return this.config.kd;
    }
    public getKs() {
        return this.config.ks;
    }
    public getFresnelModel() {
        return this.config.fresnelModel;
    }
    public getNdfModel() {
        return this.config.ndfModel;
    }
    public getGeometryModel() {
        return this.config.geometryModel;
    }

    private static clampToUnitInterval(value: number): number {
        if (value < 0.0) {
            return 0.0;
        }
        if (value > 1.0) {
            return 1.0;
        }
        return value;
    }

    /**
    Java's `MicroFacetedMaterial(String csvFileName, String materialName)`,
    over the text of the CSV rather than its file name. See the class note
    above for the split.

    @param csvText The complete contents of the microfacet CSV file
    @param materialName The `material_name` to look for, matched ignoring case
    @param sourceName The name of the read resource, for diagnostics
    */
    public static fromCsvText(csvText: string, materialName: string, sourceName: string): MicroFacetedMaterial {
        return new MicroFacetedMaterial(MicroFacetedMaterial.loadConfigFromCsvText(csvText, materialName, sourceName));
    }

    private static loadConfigFromCsvText(csvText: string, materialName: string, sourceName: string): MicrofacetConfig {
        const defaultConfig: MicrofacetConfig = new MicrofacetConfig();
        const normalizedName: string = materialName.trim().toLowerCase();
        const lines: string[] = MicroFacetedMaterial.readAsciiLines(csvText);

        const headerLine: string | null = lines.length > 0 ? lines[0]! : null;
        if (headerLine === null || headerLine.trim().length < 1) {
            throw new Error("Empty CSV header in file: " + sourceName);
        }
        const headerIndex: Map<string, number> = MicroFacetedMaterial.parseHeaderIndex(headerLine);

        let lineNumber: number;
        for (lineNumber = 1; lineNumber < lines.length; lineNumber++) {
            const line: string = lines[lineNumber]!;
            if (line.trim().length < 1) {
                continue;
            }
            const row: string[] = MicroFacetedMaterial.splitCsv(line);
            const rowName: string | null = MicroFacetedMaterial.field(row, headerIndex, "material_name");
            if (rowName === null) {
                continue;
            }
            if (rowName.trim().toLowerCase() !== normalizedName) {
                continue;
            }

            const roughness: number = MicroFacetedMaterial.clampToUnitInterval(
                MicroFacetedMaterial.doubleField(row, headerIndex, "roughness", defaultConfig.roughness),
            );
            const alpha: number = MicroFacetedMaterial.clampToUnitInterval(
                MicroFacetedMaterial.doubleField(row, headerIndex, "alpha", roughness * roughness),
            );
            const diffuseColor: ColorRgb = MicroFacetedMaterial.rgbField(
                row,
                headerIndex,
                "diffuse_r",
                "diffuse_g",
                "diffuse_b",
                defaultConfig.diffuse,
            );
            const f0: ColorRgb = MicroFacetedMaterial.rgbField(
                row,
                headerIndex,
                "f0_r",
                "f0_g",
                "f0_b",
                defaultConfig.fresnelF0,
            );
            const eta: ColorRgb = MicroFacetedMaterial.rgbField(
                row,
                headerIndex,
                "eta_r",
                "eta_g",
                "eta_b",
                defaultConfig.eta,
            );
            const kappa: ColorRgb = MicroFacetedMaterial.rgbField(
                row,
                headerIndex,
                "kappa_r",
                "kappa_g",
                "kappa_b",
                defaultConfig.kappa,
            );
            return new MicrofacetConfig(
                rowName.trim(),
                defaultConfig.ambient,
                diffuseColor,
                f0,
                defaultConfig.doubleSided,
                defaultConfig.reflectionCoefficient,
                defaultConfig.refractionCoefficient,
                defaultConfig.opacity,
                defaultConfig.phongExponent,
                roughness,
                alpha,
                f0,
                eta,
                kappa,
                MicroFacetedMaterial.clampToUnitInterval(
                    MicroFacetedMaterial.doubleField(row, headerIndex, "kd", defaultConfig.kd),
                ),
                MicroFacetedMaterial.clampToUnitInterval(
                    MicroFacetedMaterial.doubleField(row, headerIndex, "ks", defaultConfig.ks),
                ),
                MicroFacetedMaterial.intField(
                    row,
                    headerIndex,
                    "fresnel_model",
                    MicroFacetedMaterial.FRESNEL_MODEL_SCHLICK,
                ),
                MicroFacetedMaterial.intField(row, headerIndex, "ndf_model", MicroFacetedMaterial.NDF_MODEL_BECKMANN),
                MicroFacetedMaterial.intField(
                    row,
                    headerIndex,
                    "geometry_model",
                    MicroFacetedMaterial.GEOMETRY_MODEL_SMITH,
                ),
            );
        }

        throw new Error("SimpleMaterial '" + materialName + "' not found in CSV: " + sourceName);
    }

    /**
    The lines Java's loop would obtain from `readAsciiLine`: the terminator is
    dropped and a `\r` before it with it, and the segment after a final
    newline — which Java never reaches, because that read throws — is not a
    line.
    */
    private static readAsciiLines(csvText: string): string[] {
        const lines: string[] = csvText.split("\n").map((line) => line.replace(/\r/g, ""));
        if (lines.length > 0 && lines[lines.length - 1] === "") {
            lines.pop();
        }
        return lines;
    }

    private static parseHeaderIndex(headerLine: string): Map<string, number> {
        const header: string[] = MicroFacetedMaterial.splitCsv(headerLine);
        const headerIndex = new Map<string, number>();
        let i: number;
        for (i = 0; i < header.length; i++) {
            headerIndex.set(header[i]!.trim().toLowerCase(), i);
        }
        return headerIndex;
    }

    /** Java's `line.split(",", -1)`, which keeps trailing empty fields. */
    private static splitCsv(line: string): string[] {
        return line.split(",");
    }

    private static field(row: string[], headerIndex: Map<string, number>, key: string): string | null {
        const index: number | undefined = headerIndex.get(key);
        if (index === undefined) {
            return null;
        }
        if (index < 0 || index >= row.length) {
            return null;
        }
        return row[index]!.trim();
    }

    private static doubleField(
        row: string[],
        headerIndex: Map<string, number>,
        key: string,
        defaultValue: number,
    ): number {
        const value: string | null = MicroFacetedMaterial.field(row, headerIndex, key);
        if (value === null || value.trim().length < 1) {
            return defaultValue;
        }
        return Number.parseFloat(value);
    }

    private static intField(
        row: string[],
        headerIndex: Map<string, number>,
        key: string,
        defaultValue: number,
    ): number {
        const value: string | null = MicroFacetedMaterial.field(row, headerIndex, key);
        if (value === null || value.trim().length < 1) {
            return defaultValue;
        }
        return Number.parseInt(value, 10);
    }

    private static rgbField(
        row: string[],
        headerIndex: Map<string, number>,
        keyR: string,
        keyG: string,
        keyB: string,
        defaultValue: ColorRgb,
    ): ColorRgb {
        const r: number = MicroFacetedMaterial.doubleField(row, headerIndex, keyR, defaultValue.r());
        const g: number = MicroFacetedMaterial.doubleField(row, headerIndex, keyG, defaultValue.g());
        const b: number = MicroFacetedMaterial.doubleField(row, headerIndex, keyB, defaultValue.b());
        return new ColorRgb(r, g, b);
    }
}
